compiler/rustc_parse/src/validate_attr.rs

```diff
@@ -5,21 +5,68 @@ use crate::{errors, parse_in};
 use rustc_ast::token::Delimiter;
 use rustc_ast::tokenstream::DelimSpan;
 use rustc_ast::MetaItemKind;
-use rustc_ast::{self as ast, AttrArgs, AttrArgsEq, Attribute, DelimArgs, MetaItem};
+use rustc_ast::{self as ast, AttrArgs, AttrArgsEq, Attribute, DelimArgs, MetaItem, Safety};
 use rustc_errors::{Applicability, FatalError, PResult};
-use rustc_feature::{AttributeTemplate, BuiltinAttribute, BUILTIN_ATTRIBUTE_MAP};
+use rustc_feature::{AttributeSafety, AttributeTemplate, BuiltinAttribute, BUILTIN_ATTRIBUTE_MAP};
 use rustc_session::errors::report_lit_error;
-use rustc_session::lint::builtin::ILL_FORMED_ATTRIBUTE_INPUT;
+use rustc_session::lint::builtin::{ILL_FORMED_ATTRIBUTE_INPUT, UNSAFE_ATTR_OUTSIDE_UNSAFE};
 use rustc_session::lint::BuiltinLintDiag;
 use rustc_session::parse::ParseSess;
-use rustc_span::{sym, Span, Symbol};
+use rustc_span::{sym, BytePos, Span, Symbol};
 
 pub fn check_attr(psess: &ParseSess, attr: &Attribute) {
     if attr.is_doc_comment() {
         return;
     }
 
     let attr_info = attr.ident().and_then(|ident| BUILTIN_ATTRIBUTE_MAP.get(&ident.name));
+    let attr_item = attr.get_normal_item();
+
+    let is_unsafe_attr =
+        attr_info.map(|attr| attr.safety == AttributeSafety::Unsafe).unwrap_or(false);
+
+    if is_unsafe_attr {
+        if let ast::Safety::Default = attr_item.unsafety {
+            let path_span = attr_item.path.span;
+
+            // If the `attr_item`'s span is not from a macro, then just suggest
+            // wrapping it in `unsafe(...)`. Otherwise, we suggest putting the
+            // `unsafe(`, `)` right after and right before the opening and closing
+            // square bracket respectively.
+            let diag_span = if attr_item.span().can_be_used_for_suggestions() {
+                attr_item.span()
+            } else {
+                attr.span.with_lo(attr.span.lo() + BytePos(2)).with_hi(attr.span.hi() - BytePos(1))
+            };
```

Comment:
Discussion: `BytePos` manipulations always make me a bit nervous.

> right before the opening and closing square bracket respectively.

isn't 100% correct, the code is (correctly) adjusting for `#[` and `]`.

Reply:
Yeah I tried to figure out a way to not need the `BytePos` things but it doesn't seem to be possible. I can add the `unsafe(` before the path name, but the problem is in the following case, there is no span for the `$l` because the macro expansion stuff resolves it to the literal, so I can't add the `)` after.

```
macro_rules! ident2 {
    ($e:ident, $l:literal) => {
        #[$e = $l]
        //~^ ERROR: unsafe attribute used without unsafe
        //~| WARN this is accepted in the current edition
        extern fn bars() {}
    }
}
```

Reply:
> isn't 100% correct, the code is (correctly) adjusting for #[ and ].

I'm trying to say the `unsafe(...)` gets placed inside the square brackets. I can see how my comment can be confusing, but I'm not quite sure what this nit is referring to here.

Reply:
Ah I misunderstand that comment, you are correct.